### PR TITLE
General improvements

### PR DESCRIPTION
**Pull request by DeniseGiub**
This PR aims at solving the issue regarding the management of more microgrids.

## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.

### CODE REVIEW

File: scripts/add_electricity.py [hunk 1]
```diff
@@ -318,45 +348,55 @@ def attach_storageunits(n, costs, technologies, extendable_carriers):
...
-            cyclic_state_of_charge=True,
-        )
+    # Iterate through each microgrid
+    for microgrid in microgrid_ids:
+        # Iterate through each storage technology
+        for tech in technologies:
+            # Add the storage units to the power network
+            n.madd(
+                "StorageUnit",
+                {microgrid},
```
Comment (davide-f) [inline scripts/add_electricity.py:361]: when you use "{obj}", this is a set, not a list.
Moreover, you don't need the for loop to add by microgrid. you can do something like n.madd("StorageUnit", microgrid_list, ....

File: scripts/add_electricity.py [hunk 2]
```diff
@@ -318,45 +348,55 @@ def attach_storageunits(n, costs, technologies, extendable_carriers):
...
-    n_load = 1
-
-    # Create an index for the loads
-    index = pd.Index(list(range(n_load)))
-
-    # Get the index of the buses in the power network
-    buses_i = n.buses.index
+    # Rename the header of the demand df as the bus the load will be attached to
+    for i in number_microgrids:
+        load.rename(columns={f"{i+1}": f"new_bus_microgrid_{i+1}"}, inplace=True)
```
Comment (davide-f) [inline scripts/add_electricity.py:390]: This is something that I don't undestand; why is this needed?
My feeling is that the naming of the buses/micorgrids shall be more consistent and it may give problems along the chain.
The name of the bus of a microgrid shall be the name of the microgrid as specified in the config. I don't think there is the need to name them in the "hard" way inside the code.

File: scripts/build_demand.py [hunk 3]
```diff
@@ -108,25 +119,21 @@ def get_WorldPop_path(
...
+def estimate_microgrid_population(
+    p, raster_path, shapes_path, sample_profile, output_prefix, output_file
+):
+    # Read the sample profile of electricity demand and extract the column corrisponding to the electric load
+    per_unit_load = pd.read_csv(sample_profile)["0"] / p
+
+    # Dataframe of the load
+    microgrid_load = pd.DataFrame()
 
+    number_microgrids = 3  # len(os.listdir("resources/masked_files"))
```
Comment (davide-f) [inline scripts/build_demand.py:136]: This is a debug value; the number of microgrids and their name shall be provided in the config

File: scripts/create_network.py [hunk 4]
```diff
@@ -42,10 +31,84 @@ def create_network():
...
+        x, y = point_geom.x, point_geom.y
+
+        # Check for overlapping microgrids and raise an error if happening
+        if (x, y) in bus_coords:
+            raise ValueError(
+                "Overlapping microgrids detected, adjust the coordinates in the config.yaml file"
+            )
+
+        # Add the bus to the network and update the set of bus coordinates and microgrid IDs
+        n.add("Bus", bus_name, x=x, y=y, v_nom=0.220)
```
Comment (davide-f) [inline scripts/create_network.py:59]: refuse: 0.220?

File: scripts/create_network.py [hunk 5]
```diff
@@ -42,10 +31,84 @@ def create_network():
...
+    # Load the GeoJSON file
+    with open(input_file) as f:
+        data = json.load(f)
+
+    # Keep track of the bus coordinates and microgrid IDs
+    bus_coords = set()
+    microgrid_ids = set()
+
+    # Iterate over each feature in the GeoJSON file
+    for feature in data["features"]:
```
Comment (davide-f) [inline scripts/create_network.py:44]: can you explain this better?
My feeling is that this is highly time consuming

File: scripts/create_network.py [hunk 6]
```diff
@@ -42,10 +31,84 @@ def create_network():
...
+
+        # Add the edges of the triangulation to the network as lines, but only between buses in the same microgrid
+        for edge in tri.simplices:
+            if (edge[0] < len(microgrid_buses)) and (edge[1] < len(microgrid_buses)):
+                n.add(
+                    "Line",
+                    f"{microgrid_id}_line_{edge[0]}_{edge[1]}",
+                    bus0=microgrid_buses.index[edge[0]],
+                    bus1=microgrid_buses.index[edge[1]],
+                )
```
Comment (davide-f) [inline scripts/create_network.py:84]: Since you do the check before and give to the triangulation only buses belonging to the microgrid is this forloop necessary?
using n.madd would be more efficient

File: scripts/earth_osm.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+import json
+import os
+
+import matplotlib.pyplot as plt
+import numpy as np
+from _helpers_dist import configure_logging, sets_path_to_root
+from shapely.geometry import Point, Polygon
+
+
+def transform_json_to_geojson(input_file, output_file):
```
Comment (davide-f) [inline scripts/earth_osm.py:14]: use read/save geojson available in the helper, it saves you time ;)

File: scripts/earth_osm.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    if "snakemake" not in globals():
+        from _helpers_dist import mock_snakemake
+
+        os.chdir(os.path.dirname(os.path.abspath(__file__)))
+        snakemake = mock_snakemake("earth_osm")
+        sets_path_to_root("pypsa-distribution")
+
+    configure_logging(snakemake)
+
+    transform_json_to_geojson(
```
Comment (davide-f) [inline scripts/earth_osm.py:99]: So, we need a rule/script to download the raw osm data, similarly to download_osm_data in pypsa-earth and then this could be another script that filter them

File: resources/buildings/buildings.json
Comment (davide-f) [inline resources/buildings/buildings.json:1]: This file shall not be saved into the repository, delete it.
It shall be downloaded automatically using earth_osm.

**Review by davide-f — COMMENTED** [review]
Hello @DeniseGiub :)
Some preliminary comments :)
Nice job!

The problem you told me is 99% due to some naming issues and the indexing of the dataframe.
In particular, the index of the powerplants is numeric, at least the one you saved here. let's make sure that the index of the dataframe you pass to the pypsa network is consistent to make sure no problems arise

**Comment by davide-f** [comment]
@DeniseGiub use the precommit; I see a ton of pre-commits
make sure you have it installed it and when you code you have pypsa-earth enabled.
To install it do: pre-commit install

File: scripts/create_network.py (hunk 6 above)
Comment (DeniseGiub) [reply scripts/create_network.py:84]: Yes, thanks for the suggestion, I made it! :D

File: scripts/create_network.py [hunk 9]
```diff
@@ -1,29 +1,18 @@
...
-Outputs
--------
-- ``networks/base.nc``
-   
-Description
------------
-This script creates a PyPSA network with as much AC buses as microgrids specified in the file config.yaml
-"""
-
+import json
```
[the remark this answers is not in the document]
Comment (DeniseGiub) [reply scripts/create_network.py:2]: Done

File: scripts/build_demand.py [hunk 10]
```diff
@@ -77,6 +78,16 @@ def create_microgrid_shapes(microgrids_list, output_path):
         microgrid_shapes.append(microgrid_shape)
         microgrid_names.append(microgrid_name)
 
+    microgrid_gdf = gpd.GeoDataFrame(
+        {"name": microgrid_names, "geometry": microgrid_shapes}
+    )
+
+    output_dict = json.loads(microgrid_gdf.to_json())
```
[the remark this answers is not in the document]
Comment (DeniseGiub) [reply scripts/build_demand.py:85]: Yes I changed that

File: scripts/build_demand.py (hunk 3 above)
Comment (DeniseGiub) [reply scripts/build_demand.py:136]: It was an old thing, I cancelled it

File: scripts/build_demand.py [hunk 11]
```diff
@@ -145,35 +152,72 @@ def create_masked_file(raster_path, shapes_path, output_prefix):
                 }
             )
 
-        out_raster_path = f"{output_prefix}_{i+1}.tif"
+        with MemoryFile() as memfile:
+            with memfile.open(**out_meta) as dest:
+                dest.write(masked)
```
[the remark this answers is not in the document]
Comment (DeniseGiub) [reply scripts/build_demand.py:157]: Yes it was useless and senseless, thanks!

File: scripts/add_electricity.py (hunk 2 above)
Comment (DeniseGiub) [reply scripts/add_electricity.py:390]: I needed that because I want to attach the load to the central buses of each grid (each one named "new_bus_microgrid_{i}). Anyway, I made some changes in this function and in build_demand so that the process of attaching the load is much more similar to pypsa earth

File: scripts/create_network.py (hunk 4 above)
Comment (DeniseGiub) [reply scripts/create_network.py:56]: Why? Shouldn't I set a voltage level for the network?
